mrbgems/mruby-catch/mrblib/catch.rb

```diff
@@ -9,14 +9,14 @@ def initialize(tag, val)
 
 module Kernel
   def catch(tag=Object.new, &block)
-    block.call(tag)
-  rescue ThrowCatchJump => e
-    unless e._tag.equal?(tag)
-      raise e
-    end
-    return e._val
+    # A double closure is required to make the nested `catch` distinguishable
+    # and because `break` goes back to `proc->upper`.
+    -> { -> { block.call(tag) }.call }.call
```

Comment:
I am not sure I understand why double lambda is needed. Could you show me the case we need double lambda?

Reply:
My initial idea was a single lambda, but testing the nested `catch` added in #5323 revealed a problem.

After considering the cause,
 1. The `RBreak` object specifies the `proc` that issues the `break`.
 2. In the call stack unwind process, `proc-> upper` is targeted in comparison with `mrb_callinfo::proc`.  
    ref https://github.com/mruby/mruby/blob/5735d7c2fd18c7d92bfe7a7d1d68547e5765c60d/src/vm.c#L2145
 3. The `proc` object as the `catch` method does not change from the time of initialization.
 4. If `proc->upper` points to the `catch` method, it reacts to the innermost `catch`.

Therefore, `proc` and `proc->upper` need to be changed each time, which requires the generation of double lambdas.

Extra:
I mentioned that the unwind process of `break` will compare `proc->upper`, but to do this directly by comparing with `RBreak::proc`, it need to change the VM.
I think that probably leads to simplification of the exception handling part, but I haven't succeeded.
If all goes well, I think `RBREAK_TAG` will be combined into `RBREAK_TAG_BREAK`, `RBREAK_TAG_JUMP` and `RBREAK_TAG_STOP`.
It's not working well so far, so it's a mochi that I drew in a picture.

Reply:
OK. When I tried single lambda worked for nested catches as well. I don't know what caused the difference.
In any way I updated `mruby-catch` but I left double lambda unchanged.

Reply:
I tried single lambda again, but it still fails `rake test`.

After thinking about it, I realized that you might have kept `ci[2].proc` as is.
https://github.com/mruby/mruby/blob/d178e147f5cd86af247fc4b6cfafbde1308ede33/mrbgems/mruby-catch/src/mruby-catch.c#L98

If that guess is correct, it can pass the `rake test`, but it seems to cause another problem.

```console
% cat catch1.rb
def xxx
  proc { throw 1 }
end

catch(1, &xxx)

% bin/mruby catch1.rb
trace (most recent call last):
catch1.rb:5: break from proc-closure (LocalJumpError)
```

Reply:
OK, I will keep double lambda. Thank you!